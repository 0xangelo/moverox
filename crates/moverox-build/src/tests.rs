use std::path::Path;

use itertools::Itertools as _;
use testresult::TestResult;

use crate::move_package;

#[test]
fn generate_rust_for_move_stdlib() -> TestResult {
    let pkg_path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("move")
        .join("move-stdlib");

    let builder = move_package(pkg_path, "MoveStdlib").published_at("0x1");
    let move_files = {
        let mut paths = builder.collect_move_files()?;
        paths.sort();
        paths
    };

    let files_found = move_files
        .iter()
        .map(|path| path.file_name().unwrap().display())
        .join("\n");
    insta::assert_snapshot!(files_found, @r"
    address.move
    ascii.move
    bcs.move
    bit_vector.move
    bool.move
    debug.move
    fixed_point32.move
    hash.move
    macros.move
    option.move
    string.move
    type_name.move
    u128.move
    u16.move
    u256.move
    u32.move
    u64.move
    u8.move
    unit_test.move
    uq32_32.move
    uq64_64.move
    vector.move
    ");

    let modules = builder.parse_files(&move_files)?;
    let rust_code = builder.generate_rust_str(&modules)?;

    let move_stdlib = prettyplease::unparse(&syn::parse_file(&rust_code).unwrap());
    insta::assert_snapshot!("MoveStdLib", move_stdlib);
    Ok(())
}

#[test]
fn generate_rust_for_sui_framework() -> TestResult {
    let pkg_path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("move")
        .join("sui-framework");

    let builder = move_package(pkg_path, "Sui")
        .published_at("0x2")
        .with_implicit_sui_imports()
        .map_address("std", "::moverox_sui::std");
    let move_files = {
        let mut paths = builder.collect_move_files()?;
        paths.sort();
        paths
    };

    let files_found = move_files
        .iter()
        .map(|path| path.file_name().unwrap().display())
        .join("\n");
    insta::assert_snapshot!(files_found, @r"
    accumulator.move
    address.move
    authenticator_state.move
    bag.move
    balance.move
    bcs.move
    borrow.move
    clock.move
    coin.move
    config.move
    deny_list.move
    display.move
    dynamic_field.move
    dynamic_object_field.move
    event.move
    hex.move
    linked_table.move
    math.move
    object.move
    object_bag.move
    object_table.move
    package.move
    party.move
    pay.move
    priority_queue.move
    prover.move
    random.move
    sui.move
    table.move
    table_vec.move
    token.move
    transfer.move
    tx_context.move
    types.move
    url.move
    vec_map.move
    vec_set.move
    versioned.move
    ");

    let modules = builder.parse_files(&move_files)?;
    let rust_code = builder.generate_rust_str(&modules)?;

    let sui_framework = prettyplease::unparse(&syn::parse_file(&rust_code).unwrap());
    insta::assert_snapshot!("Sui", sui_framework);
    Ok(())
}
