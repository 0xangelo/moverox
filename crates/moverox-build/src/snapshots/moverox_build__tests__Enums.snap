---
source: crates/moverox-build/src/tests.rs
expression: enums
---
#[allow(rustdoc::all)]
#[cfg(not(doctest))]
pub mod enums {
    #[allow(non_camel_case_types, unused)]
    type address = ::moverox::types::Address;
    #[allow(non_camel_case_types, unused)]
    type u256 = ::moverox::types::U256;
    #[allow(non_camel_case_types, unused)]
    type vector<T> = ::std::vec::Vec<T>;
    #[derive(
        Clone,
        Debug,
        PartialEq,
        Eq,
        Hash,
        ::moverox::traits::MoveDatatype,
        ::moverox::serde::Deserialize,
        ::moverox::serde::Serialize,
    )]
    #[move_(crate = ::moverox::traits)]
    #[serde(crate = "::moverox::serde")]
    #[move_(module = enums)]
    #[allow(non_snake_case)]
    pub enum Single {
        Only,
    }
    /// `Segment` enum definition.
    /// Defines various string segments.
    #[derive(
        Clone,
        Debug,
        PartialEq,
        Eq,
        Hash,
        ::moverox::traits::MoveDatatype,
        ::moverox::serde::Deserialize,
        ::moverox::serde::Serialize,
    )]
    #[move_(crate = ::moverox::traits)]
    #[serde(crate = "::moverox::serde")]
    #[move_(module = enums)]
    #[allow(non_snake_case)]
    pub enum Segment {
        /// Empty variant, no value.
        Empty,
        /// Variant with a value (positional style).
        String(std::string::String),
        /// Variant with named fields.
        Special { content: vector<u8>, encoding: u8 },
    }
    #[derive(
        Clone,
        Debug,
        PartialEq,
        Eq,
        Hash,
        ::moverox::traits::MoveDatatype,
        ::moverox::serde::Deserialize,
        ::moverox::serde::Serialize,
    )]
    #[move_(crate = ::moverox::traits)]
    #[serde(crate = "::moverox::serde")]
    #[move_(module = enums)]
    #[allow(non_snake_case)]
    pub enum Generic<T> {
        Unit(#[serde(skip)] ::std::marker::PhantomData<T>),
        Tuple(u64),
        Struct { value: u64 },
    }
}
#[allow(rustdoc::all)]
#[cfg(not(doctest))]
pub mod other {
    #[allow(non_camel_case_types, unused)]
    type address = ::moverox::types::Address;
    #[allow(non_camel_case_types, unused)]
    type u256 = ::moverox::types::U256;
    #[allow(non_camel_case_types, unused)]
    type vector<T> = ::std::vec::Vec<T>;
    #[derive(
        Clone,
        Debug,
        PartialEq,
        Eq,
        Hash,
        ::moverox::traits::MoveDatatype,
        ::moverox::serde::Deserialize,
        ::moverox::serde::Serialize,
    )]
    #[move_(crate = ::moverox::traits)]
    #[serde(crate = "::moverox::serde")]
    #[move_(module = other)]
    #[allow(non_snake_case)]
    pub enum Wrapper<T> {
        Single(super::enums::Single),
        OtherPositional(T),
        OtherNamed { inner: T },
    }
    /// Type parameter `Single` should shadow the imported type with the same name.
    #[derive(
        Clone,
        Debug,
        PartialEq,
        Eq,
        Hash,
        ::moverox::traits::MoveDatatype,
        ::moverox::serde::Deserialize,
        ::moverox::serde::Serialize,
    )]
    #[move_(crate = ::moverox::traits)]
    #[serde(crate = "::moverox::serde")]
    #[move_(module = other)]
    #[allow(non_snake_case)]
    pub enum Shadowed<Single> {
        Only(Single),
    }
}
